src/lib.rs

```diff
@@ -151,40 +152,80 @@ struct HelperInner {
     consumer_done: bool,
 }
 
-/// Error type for `from_env` function.
+/// Return type for `from_env_ext` function.
 #[derive(Debug)]
-pub enum ErrFromEnv {
-    /// There isn't env var, that describes jobserver to inherit.
-    IsNotConfigured,
-    /// Cannot connect following this process's environment.
-    PlatformSpecific {
-        /// Error.
-        err: io::Error,
-        /// Name of gotten env var.
-        env: &'static str,
-        /// Value of gotten env var.
-        var: String,
-    },
+pub struct FromEnv {
+    /// Result of trying to get jobserver client from env.
+    pub client: Result<Client, FromEnvError>,
+    /// Name of the environment variable.
+    /// Empty if no relevant environment variable is found.
+    pub var_name: &'static str,
+    /// Value of the environment variable.
+    /// Empty if no relevant environment variable is found.
+    pub var_value: OsString,
 }
 
-impl std::fmt::Display for ErrFromEnv {
+/// Error type for `from_env_ext` function.
+#[derive(Debug)]
+#[non_exhaustive]
+pub enum FromEnvError {
+    /// There is no environment variable that describes jobserver to inherit.
+    NoEnvVar,
+    /// Cannot parse jobserver environment variable value, incorrect format.
+    CannotParse(String),
+    /// Cannot open path or name from the jobserver environment variable value.
+    CannotOpenPath(String, io::Error),
+    /// Cannot open file descriptor from the jobserver environment variable value.
+    CannotOpenFd(std::ffi::c_int, io::Error),
```

Comment:
Is it really ok to expose unix-specific error variant?

```suggestion
    #[cfg(unix)]
    CannotOpenFd(std::os::fd::RawFd, io::Error),
```

Reply:
Code using jobserver-rs will be target-independent, most likely, so it's probably simpler to avoid cfgs.

Reply:
Even so, I think it's better to use std::os::fd::RawFd instead of c_int

Reply:
Thanks, it's a shame that we have to use a raw integer type here and expose what I considered to be implementation detail, but I can't think of something better right now.

Reply:
@petrochenkov Could we merge CannotOpen{Path, Fd} into one and make the descriptor (path/fd) opaque?

I don't think cargo would go to the extent of inspecting which path/fd that open failed, so maybe we don't need to expose such impl detail and os specific detail?

Reply:
> But it could probably be solved by adding a new `FromEnvError` variant too.

IMO adding a new variant is better than exposing implementation detail and let the user check it themselves.

Reply:
If fd is going to be used only for error message, would it be enough to note is it write or read descriptor, without actual number that is in environment variable value? imho negative file descriptors and things like that should be handled within this crate.

Reply:
If people are not against multiplying public types, then we can go with
```rust
pub enum FromEnvError {
    NoEnvVar,
    CannotParse(CannotParseError),
    CannotOpenPath(CannotOpenPathError),
    CannotOpenFd(CannotOpenFdError),
    NotAPipe(NotAPipeError),
    Unsupported,
}
```
where `{CannotParse,CannotOpenPath,CannotOpenFd,NotAPipe}Error` are all opaque structures.

This would be future-compatible with keeping all kind of stuff inside them.

Reply:
>where `{CannotParse,CannotOpenPath,CannotOpenFd,NotAPipe}Error` are all opaque structures

They can even be made unnameable for a start.
```
mod error  {
    pub struct CannotParseError { ... }
}

pub enum FromEnvError {
    CannotParse(CannotParseError),
    ...
}
```

Reply:
If that's the case, why not make `FromEnvError` and opaque type and only expose:

```rust
pub enum FromEnvErrorKind {
    CannotParse,
    CannotOpenPath,
    CannotOpenFd,
    NotAPipe,
}

impl FromEnvError {
    pub fn kind(&self) -> FromEnvErrorKind;
}
```

Reply:
Yeah, that should be fine too.